src/System.Device.Gpio.Tests/GpioControllerTestBase.cs

```diff
@@ -153,7 +150,81 @@ void callback(object sender, PinValueChangedEventArgs pinValueChangedEventArgs)
             }
         }
 
+        [Fact]
+        public void AddCallbackRemoveCallbackTest()
+        {
+            int risingEventOccuredCount = 0, fallingEventOccuredCount = 0;
+            using (GpioController controller = new GpioController(GetTestNumberingScheme(), GetTestDriver()))
+            {
+                controller.OpenPin(InputPin, PinMode.Input);
+                controller.OpenPin(OutputPin, PinMode.Output);
+                controller.Write(OutputPin, PinValue.Low);
+
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Rising, (o, e) => {
+                    risingEventOccuredCount++;
+                });
+                                
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Rising, callback);
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Rising, (o, e) => {
+                    risingEventOccuredCount++;
+                    if (fallingEventOccuredCount == 4)
+                    {
+                        controller.UnregisterCallbackForPinValueChangedEvent(InputPin, callback);
+                    }
+                });
+
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Falling, (o, e) => {
+                    fallingEventOccuredCount++;
+                });
+
+                for (int i = 0; i < 10; i++)
+                {
+                    controller.Write(OutputPin, PinValue.High);
+                    Thread.Sleep(10);
+                    controller.Write(OutputPin, PinValue.Low);
+                    Thread.Sleep(10);
+                }
+                Assert.Equal(25, risingEventOccuredCount);
+                Assert.Equal(10, fallingEventOccuredCount);
+
+                void callback(object sender, PinValueChangedEventArgs e)
+                {
+                    risingEventOccuredCount++;
+                }
+            }
+        }
+
+        [Fact]
+        public void WaitForEventTest()
+        {
+            using (GpioController controller = new GpioController(GetTestNumberingScheme(), GetTestDriver()))
+            {
+                bool cancelled = false;
+                CancellationTokenSource tokenSource = new CancellationTokenSource();
+                controller.OpenPin(InputPin, PinMode.Input);
+                controller.OpenPin(LedPin, PinMode.Input);
+                controller.OpenPin(OutputPin, PinMode.Output);
+                controller.Write(OutputPin, PinValue.Low);
+
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Rising, (o, e) =>
+                {
+                        tokenSource.Cancel();
+                        cancelled = true;
+                });
+
+                Task.Run(() =>
+                {
+                    controller.Write(OutputPin, PinValue.High);
+                    Thread.Sleep(10);
```

Comment:
what's the setting to low for? You seem to be waiting only for raising

Reply:
just for saving energy 😛

Reply:
removed